File: python/63. unique-paths-ii.py
from typing import List


# lc sol start
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        r, c = len(obstacleGrid), len(obstacleGrid[0])

        obstacleGrid[0][0] = 1 - obstacleGrid[0][0]

        for i in range(1, r):
            obstacleGrid[i][0] = obstacleGrid[i - 1][0] * (1 - obstacleGrid[i][0])

        for i in range(1, c):
            obstacleGrid[0][i] = obstacleGrid[0][i - 1] * (1 - obstacleGrid[0][i])

        for i in range(1, r):
            for j in range(1, c):
                obstacleGrid[i][j] = (obstacleGrid[i - 1][j] + obstacleGrid[i][j - 1]) * (1 - obstacleGrid[i][j])

        return obstacleGrid[-1][-1]


# lc sol end


def main() -> None:
    test_cases = {
        'inputs': [([[0, 0, 0], [0, 1, 0], [0, 0, 0]],), ([[0, 1], [0, 0]],)],
        'outputs': [2, 1],
    }

    solution = Solution()

    for idx, args in enumerate(test_cases['inputs']):
        op = solution.uniquePathsWithObstacles(*args)
        assert op == test_cases['outputs'][idx], \
            f"expected output {test_cases['outputs'][idx]} for input {args}, got {op}"

    print('all sample test cases passed successfully!')


if __name__ == '__main__':
    main()

File: cpp/63. unique-paths-ii.cpp
//
// Created by Rutvik Patel on 1/31/24.
//

#include "include.h"

// solution
class Solution {
public:
    int uniquePathsWithObstacles(std::vector<std::vector<int>>& obstacle_grid) {
        int n_rows = static_cast<int>(obstacle_grid.size()), n_cols = static_cast<int>(obstacle_grid[0].size());
        obstacle_grid[0][0] = 1 - obstacle_grid[0][0];

        for (int idx = 1; idx < n_rows; ++idx)
            obstacle_grid[idx][0] = obstacle_grid[idx - 1][0] * (1 - obstacle_grid[idx][0]);

        for (int idx = 1; idx < n_cols; ++idx)
            obstacle_grid[0][idx] = obstacle_grid[0][idx - 1] * (1 - obstacle_grid[0][idx]);

        for (int i = 1; i < n_rows; ++i)
            for (int j = 1; j < n_cols; ++j)
                obstacle_grid[i][j] = (obstacle_grid[i - 1][j] + obstacle_grid[i][j - 1]) * (1 - obstacle_grid[i][j]);

        return obstacle_grid.back().back();
    }
};

// main
int main(__attribute__((unused)) int argc, __attribute__((unused)) char *argv[]) {
    std::vector<std::pair<std::vector<std::vector<int>>, int>> test_cases = {
        { { { 0, 0, 0 }, { 0, 1, 0 }, { 0, 0, 0} }, 2 },
        { { { 0, 1 }, { 0, 0 } }, 1 },
    };

    Solution solution;

    for (auto& [args, ans] : test_cases)
        assert_m(solution.uniquePathsWithObstacles(args) == ans, "Failed!");

    log("All testcases passed!");
}